network/src/testutils/test_node.rs

```diff
@@ -0,0 +1,313 @@
+};
+use diem_types::PeerId;
+use futures::StreamExt;
+use netcore::transport::ConnectionOrigin;
+use std::{collections::HashMap, sync::Arc};
+
+pub type InboundMessageSender =
+    channel::diem_channel::Sender<(PeerId, ProtocolId), PeerManagerNotification>;
+pub type ConnectionUpdateSender = crate::peer_manager::conn_notifs_channel::Sender;
+pub type OutboundMessageReceiver =
+    channel::diem_channel::Receiver<(PeerId, ProtocolId), PeerManagerRequest>;
+
+/// A connection handle describing the network for a node
+/// Use this to interact with the node
+#[derive(Clone)]
+pub struct InboundNetworkHandle {
+    /// To send new incoming network messages
+    pub inbound_message_sender: InboundMessageSender,
+    /// To send new incoming connections or disconnections
+    pub connection_update_sender: ConnectionUpdateSender,
+    /// To update the local state (normally done by peer manager)
+    pub peer_metadata_storage: Arc<PeerMetadataStorage>,
+}
+
+impl InboundNetworkHandle {
+    /// Push connection update, and update the local storage
+    pub fn connect(
+        &self,
+        self_peer_id: PeerId,
+        network_id: NetworkId,
+        conn_metadata: ConnectionMetadata,
+    ) {
+        // PeerManager pushes this data before it's received by events
+        self.peer_metadata_storage
+            .insert_connection(network_id, conn_metadata.clone());
+        self.connection_update_sender
+            .push(
+                conn_metadata.remote_peer_id,
+                ConnectionNotification::NewPeer(
+                    conn_metadata,
+                    // TODO will `RoleType` matter
+                    NetworkContext::new(RoleType::Validator, network_id, self_peer_id),
+                ),
+            )
+            .unwrap();
+    }
+
+    /// Push disconnect update, and update the local storage
+    pub fn disconnect(
+        &self,
+        self_peer_id: PeerId,
+        network_id: NetworkId,
+        conn_metadata: ConnectionMetadata,
+    ) {
+        // PeerManager pushes this data before it's received by events
+        self.peer_metadata_storage.remove(&PeerNetworkId::new(
+            network_id,
+            conn_metadata.remote_peer_id,
+        ));
+        self.connection_update_sender
+            .push(
+                conn_metadata.remote_peer_id,
+                ConnectionNotification::LostPeer(
+                    conn_metadata,
+                    // TODO will `RoleType` matter
+                    NetworkContext::new(RoleType::Validator, network_id, self_peer_id),
+                    DisconnectReason::ConnectionLost,
+                ),
+            )
+            .unwrap();
+    }
+}
+
+/// An application specific network handle
+pub type ApplicationNetworkHandle<Sender, Events> = (NetworkId, Sender, Events);
+
+/// A unique identifier of a node across the entire network
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct NodeId {
+    pub owner: u32,
```

Comment:
So, it originally was going to be like the actual network, where the validators & VFNs have the same PeerId as their account, etc.  However, it made things overly complicated, so I simplified it to a number.  This allows us to just have a unique identifier, and at a later time if we want to autoconnect a network (you can, since owner is same for VFN and validator).

Those were the extra flaky tests and they weren't so happy with RPC, so I haven't implemented them again yet.